src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+ using the _conceptual_ algorithm, described below (aka "tetris algorithm"). The actual implementation of layering might
+ be done differently, as long as the resulting layering is the same as if running the conceptual algorithm.
+
+1. The tracer will have a setting for preferred layer duration: P.
+1. The first encountered operation of __non-zero__ duration N is added into layer L(0, max(P,N)). The value
+ for _conditional barrier_ is set to 0.
+1. When conditional callback is encountered, the layer L(t,N) of the measurement that produced the result the conditional
+ is dependent on, is looked up and the _conditional barrier_ is set to _t + N_. At the end of the conditional scope the
+ barrier is reset to 0. (Effectively, no operations, conditioned on the result of a measurement, can happen before or in
+ the same layer as the measurement, even if they don't involve the measured qubits.)
+1. Suppose, there are already layers L(0,N0), ... , L(k,Nk) and the operation being executed is a single-qubit _op_ of
+ duration __0__ (controlled and multi-qubit operations of duration 0 are treated the same as non-zero operations).
+ Starting at L(k, Nk) and scanning backwards to L(conditional barrier, Nb) find the _first_ layer that contains an
+ operation that acts on at least one of the qubits _op_ is acting on. Add _op_ into this layer. If no such layer found,
+ remember that the qubits have pending 0-duration _op_ and commit it to the layer that first uses any of the qubits in a
+ an operation of non-zero duration. At the end of the program commit all pending operations of duration zero into a new
+ layer.
+1. Suppose, there are already layers L(0,N0), ... , L(k,Nk) and the operation being executed is _op_ of duration _N > 0_.
+ Starting at L(k, Nk) and scanning backwards to L(conditional barrier, Nb) find the _last_ layer L(t, Nt) such that
+ Qubits(t, Nt) don't contain any of the _op_'s qubits and find the _first_ layer L(w, Nw) such that Qubits(w, Nw) contains
+ some of _op_'s qubits but Nw + N <= P. Add _op_ into one of the two layer with later time. If neither such layers found,
+ add _op_ into a new layer L(k+1, max(P, N)).
+
+## Special handling of SWAP ##
+
+The tracer will provide a way to handle SWAP as, effectively, renaming of the involved qubits. The client will have the
+ choice of using the special handling versus treating the gate as a standard counted intrinsic.
+
+## Frame tracking ##
+
+The client might want to count differently operations that are applied in a different state. For example, if Hadamard gate
+ is applied to a qubit and then Rz and Mx gates, the client might want to count the sequence as if Rz as Mz were executed.
+ The frame is closed when the state of the qubit is reset (in Hadamard's case, another Hadamard operator is applied to
+ the qubit). The client will be able to register the required frame tracking with the tracer via a C++ registration
+ callback.

```

Comment:
> For example, if Hadamard gate is applied to a qubit and then Rz and Mx gates, the client might want to count the sequence as if Rz as Mz were executed.

I don't quite get the point of this sentence. Do you mean counting Rz and Mx _only after Hadamard gate_?

Reply:
within the frame, Rx is counted as Rz. added "instead" at the end of the sentence